File: packages/react-native/ReactAndroid/src/main/java/com/facebook/react/modules/network/ReactCookieJarContainer.java
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

package com.facebook.react.modules.network;

import androidx.annotation.Nullable;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import okhttp3.Cookie;
import okhttp3.CookieJar;
import okhttp3.Headers;
import okhttp3.HttpUrl;

/** Basic okhttp3 CookieJar container */
public class ReactCookieJarContainer implements CookieJarContainer {

  @Nullable private CookieJar cookieJar = null;

  @Override
  public void setCookieJar(CookieJar cookieJar) {
    this.cookieJar = cookieJar;
  }

  @Override
  public void removeCookieJar() {
    this.cookieJar = null;
  }

  @Override
  public void saveFromResponse(HttpUrl url, List<Cookie> cookies) {
    if (cookieJar != null) {
      cookieJar.saveFromResponse(url, cookies);
    }
  }

  @Override
  public List<Cookie> loadForRequest(HttpUrl url) {
    if (cookieJar != null) {
      List<Cookie> cookies = cookieJar.loadForRequest(url);
      ArrayList<Cookie> validatedCookies = new ArrayList<>();
      for (Cookie cookie : cookies) {
        try {
          Headers.Builder cookieChecker = new Headers.Builder();
          cookieChecker.add(cookie.name(), cookie.value());
          validatedCookies.add(cookie);
        } catch (IllegalArgumentException ignored) {
        }
      }
      return validatedCookies;
    }
    return Collections.emptyList();
  }
}

File: packages/react-native/ReactAndroid/src/main/java/com/facebook/react/modules/network/ReactCookieJarContainer.kt
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

@file:Suppress("DEPRECATION_ERROR") // Conflicting okhttp versions

package com.facebook.react.modules.network

import java.util.ArrayList
import okhttp3.Cookie
import okhttp3.CookieJar
import okhttp3.Headers
import okhttp3.HttpUrl

/** Basic okhttp3 CookieJar container */
public class ReactCookieJarContainer : CookieJarContainer {

  private var cookieJar: CookieJar? = null

  override fun setCookieJar(cookieJar: CookieJar) {
    this.cookieJar = cookieJar
  }

  override fun removeCookieJar() {
    cookieJar = null
  }

  override fun saveFromResponse(url: HttpUrl, cookies: List<Cookie>) {
    cookieJar?.saveFromResponse(url, cookies)
  }

  override fun loadForRequest(url: HttpUrl): List<Cookie> {
    val cookieJar = cookieJar ?: return emptyList()
    val cookies = cookieJar.loadForRequest(url)
    val validatedCookies = ArrayList<Cookie>()
    for (cookie in cookies) {
      try {
        val cookieChecker = Headers.Builder()
        cookieChecker.add(cookie.name(), cookie.value())
        validatedCookies.add(cookie)
      } catch (ignored: IllegalArgumentException) {}
    }
    return validatedCookies
  }
}
